BE#31: Password Endpoints

SoySauceLabs opened the pull request:
# What To Test:

## Using Postman - test the endpoints with the content in the doc repo:
- [ ] **Project Documentation > Wiki+Reference > BE > Password Endpoint Tests**

## Expected Results:
First...
- [ ] Log in to the DB to make observations in the **user** table.

Then...
- [ ] Observe password changes on **/password/change** endpoint using requests from **Password Endpoint Tests**.
- [ ] Note the _before_ password (in the DB) for the test user in **Password Endpoint Tests** for the **/password/reset** endpoint.
- [ ] Send request from **Password Endpoint Tests** the **/password/reset** endpoint.
- [ ]  Log in to email account listed in **Project Documentation > Wiki+Reference > BE > Password Reset Email**.
- [ ] Observe new email in inbox with password reset link.
- [ ] Observe password changes (in the DB) for the test user in **Password Endpoint Tests** for  **/password/password_rest** endpoint using request from **Password Endpoint Tests**

scotte216 commented on src/cassy.py:
```diff
@@ -205,6 +206,37 @@ def get_user_password(username):
+
+    try:
+        if username == '':
+            raise PlantalyticsLoginException(EMAIL_ERROR)
```
The EMAIL_ERROR doesn't necessarily seem like a LoginException. Although there is no functional difference between our PlantalyticExceptions, they're there for book-keeping purposes so it might be useful to create a new one. Or not.

scotte216 commented on src/cassy.py:
```diff
@@ -281,9 +313,72 @@ def verify_auth_token(auth_token):
+
+def change_user_password(username, new_password, old_password):
+    """
+    Chanes current password of the supplied username
```
Typo: Chanes should be Changes

scotte216 commented on src/password/views.py:
```diff
@@ -0,0 +1,214 @@
+                raise PlantalyticsLoginException(LOGIN_ERROR)
+            if new_password == stored_password:
+                logger.warn('Invalid new password.')
+                raise PlantalyticsLoginException(LOGIN_ERROR)
```
I think here a more descriptive error would be appropriate. Contrast to the previous exception where the username doesn't match (403 -- Forbidden, generic login-error), here they are the appropriate user, just don't have a good new password. (400 -- bad request, bad password choice).

scotte216 commented on src/password/views.py:
```diff
@@ -0,0 +1,214 @@
+        new_password_object = {}
+        new_password_object['new_password'] = new_password
+
+        return HttpResponse(json.dumps(new_password_object), content_type='application/json')
```
This might be useful to testing, but I doubt we'll want to return this object.

scotte216 commented on src/password/views.py:
```diff
@@ -0,0 +1,214 @@
+            fail_silently=False,
+        )
+
+        return HttpResponse(json.dumps(reset_token_object), content_type='application/json')
```
I don't think we want to return the token in the response.

scotte216 commented on src/password/tests.py:
```diff
@@ -0,0 +1,335 @@
+        """
+        setup_test_environment()
+        client = Client()
+        response = client.get(
```
This should be a post, not get.

SoySauceLabs replied:
This test is intended to check for an incorrect HTTP method call. It's been updated and a more appropriate response code is used now (405).

scotte216 commented on src/password/tests.py:
```diff
@@ -0,0 +1,335 @@
+            + '&old=badpass'
+            + '&password=newpass'
+        )
+        self.assertEqual(response.status_code, 400)
```
When the response code is an error, should also check the error code to be sure it's what we expect. See lines 163/164 for example.

SoySauceLabs replied:
In this case, there is no error code. However, all tests with error codes now have appropriate assertions added.

scotte216 commented on src/password/tests.py:
```diff
@@ -0,0 +1,335 @@
+            data=json.dumps(body),
+            content_type='application/json'
+        )
+        self.assertEqual(response.status_code, 403)
```
check error and expect 400 -- bad request

scotte216 commented on src/password/tests.py:
```diff
@@ -0,0 +1,335 @@
+            data=json.dumps(body),
+            content_type='application/json'
+        )
+        self.assertEqual(response.status_code, 403)
```
expect 400

scotte216 commented on src/password/tests.py:
```diff
@@ -0,0 +1,335 @@
+            content_type='application/json'
+        )
+        cassy_auth.assert_called_once_with(auth_token)
+        self.assertEqual(response.status_code, 403)
```
expect 400. Also cassy_mock/@patch cassy.change_user_password are unused.

scotte216 commented on src/password/tests.py:
```diff
@@ -0,0 +1,335 @@
+        )
+        response_body = json.loads(reset_response.content.decode('utf-8'))
+        reset_token = response_body.get('reset_token', '')
+        final_response = client.get(
```
If we don't return the token in the response (my recommendation), we can still test this with a mock and assert send_mail() was called with what we expect.

SoySauceLabs commented:
Alllllllll righty then.....

I'm pretty sure I've addressed all of the comments by making the appropriate changes (and in the cases where I have not made suggest change, I've added a line note to indicate why). Hopefully I covered everything, I forgot that the line comments get all messed up in github after you commit changes 😢 .

Annnndddd we've got more tests. I really started squeezing the last molecules of lemon juice (with a sledgehammer) to cover more password-related lines of code! 🔨 

Oh, and once thing that wasn't explicitly commented on, but discussed on Slack: The random password generation had been removed (but repurposed in the unit tests! haha!) 💃
